opt: inline all filters with virtual columns

mgartner opened the pull request:
#### opt: inline all filters with virtual columns

Previously, `InlineSelectVirtualColumns` inlined virtual column
expressions in filters and pushed those filters below a Project only if
the virtual columns were indexed. Now, all virtual column expressions
are inlined. This allows exploration rules to generate expressions that
utilize partial indexes with predicates that reference virtual columns.

We considered extending `InlineSelectVirtualColumns` to also inline a
virtual columns if they are referenced in partial index predicates.
This would require addition metadata to keep track of the set of such
virtual columns. Virtual columns have already been inlined in the
predicates in the table metadata, so simply fetching the outer columns
of those predicates would not suffice. Because this rule already is
highly selective (only operating on virtual columns), we opted to
simplify the rule to inline all virtual columns instead.

There is no release note because virtual columns are gated behind an
experimental session setting.

Release note: None

#### opt: fix optsteps panic from InlineSelectVirtualColumns

Previously, running the `optsteps` test command panicked with "could not
find path to expr" for queries when `InlineSelectVirtualColumns` was
matched. This commit fixes the panic by eliminating the anti-pattern of
calling a constructor during the pattern-match phase of a normalization
rule.

The `optsteps` test command displays step `n` of optimization by running
the optimizer twice and halting after `n` and `n - 1` rules have been
applied. The difference in the two resulting expressions represents the
change in step `n`. `InlineSelectVirtualColumns` previously called
`ConstructFiltersItem` during the pattern-match phase of the rule, but
only added the constructed filters item to the memo in the replace
phase of the rule. When some normalization rule applied during the call
to `ConstructFiltersItem` was the last rule applied in a step, the
constructed item was not added to the memo because the limit on the
number of rules prevented the replace phase of
`InlineSelectVirtualColumns`. As a result, `optsteps` panicked when it
could not find the normalized `FiltersItem` in the memo.

Release note: None

mgartner committed:
opt: inline all filters with virtual columns

Previously, `InlineSelectVirtualColumns` inlined virtual column
expressions in filters and pushed those filters below a Project only if
the virtual columns were indexed. Now, all virtual column expressions
are inlined. This allows exploration rules to generate expressions that
utilize partial indexes with predicates that reference virtual columns.

We considered extending `InlineSelectVirtualColumns` to also inline a
virtual columns if they are referenced in partial index predicates.
This would require addition metadata to keep track of the set of such
virtual columns. Virtual columns have already been inlined in the
predicates in the table metadata, so simply fetching the outer columns
of those predicates would not suffice. Because this rule already is
highly selective (only operating on virtual columns), we opted to
simplify the rule to inline all virtual columns instead.

There is no release note because virtual columns are gated behind an
experimental session setting.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60553)
<!-- Reviewable:end -->

mgartner committed:
opt: fix optsteps panic from InlineSelectVirtualColumns

Previously, running the `optsteps` test command panicked with "could not
find path to expr" for queries when `InlineSelectVirtualColumns` was
matched. This commit fixes the panic by eliminating the anti-pattern of
calling a constructor during the pattern-match phase of a normalization
rule.

The `optsteps` test command displays step `n` of optimization by running
the optimizer twice and halting after `n` and `n - 1` rules have been
applied. The difference in the two resulting expressions represents the
change in step `n`. `InlineSelectVirtualColumns` previously called
`ConstructFiltersItem` during the pattern-match phase of the rule, but
only added the constructed filters item to the memo in the replace
phase of the rule. When some normalization rule applied during the call
to `ConstructFiltersItem` was the last rule applied in a step, the
constructed item was not added to the memo because the limit on the
number of rules prevented the replace phase of
`InlineSelectVirtualColumns`. As a result, `optsteps` panicked when it
could not find the normalized `FiltersItem` in the memo.

Release note: None

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-:-MTfkfAAANLnEprTAETs:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-:-MTfnJSqBE2avjRcdSfA:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/norm/inline_funcs.go, line 244 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTflci32iX-YEdcUdyU:-MTflci32iX-YEdcUdyV:bb7o9pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cccb58889f4eca09922aeefb9e712e028fa2285/pkg/sql/opt/norm/inline_funcs.go#L244)):*
> ```Go
> 		}
> 
> 		// Initialize inlined lazily.
> ```

inlined -> inlinableFilters
___
*[pkg/sql/opt/norm/rules/inline.opt, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTflOXh6a4fmljjhYxe:-MTflOXh6a4fmljjhYxf:b-oanu3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cccb58889f4eca09922aeefb9e712e028fa2285/pkg/sql/opt/norm/rules/inline.opt#L189)):*
> ```
> #     a INT,
> #     b INT,
> #     v INT AS (abs(a)),
> ```

should this have the tag `VIRTUAL`?
___
*[pkg/sql/opt/xform/testdata/rules/select, line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTfktww3xLJ43UaZ5vs:-MTfktww3xLJ43UaZ5vt:b-p5q0yp) ([raw file](https://github.com/cockroachdb/cockroach/blob/586f4749318ef2b1d2de0477e367f0d4947e430a/pkg/sql/opt/xform/testdata/rules/select#L1735)):*
> ```
>            └── b:3 = 10 [outer=(3), constraints=(/3: [/10 - /10]; tight), fd=()-->(3)]
> 
> # Test that we can constrain an partial index with a virtual column in the
> ```

[nit] an partial -> a partial


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60553)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @rytaft)

___
*[pkg/sql/opt/norm/inline_funcs.go, line 244 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTflci32iX-YEdcUdyU:-MTgBc5qDkM1KxheNTB4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cccb58889f4eca09922aeefb9e712e028fa2285/pkg/sql/opt/norm/inline_funcs.go#L244)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

inlined -> inlinableFilters
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/inline.opt, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTflOXh6a4fmljjhYxe:-MTgBsTE4hq0jCjcTior:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cccb58889f4eca09922aeefb9e712e028fa2285/pkg/sql/opt/norm/rules/inline.opt#L189)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should this have the tag `VIRTUAL`?
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/testdata/rules/select, line 1735 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60553#-MTfktww3xLJ43UaZ5vs:-MTgC2eC-E4Lb5ymd0-J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/586f4749318ef2b1d2de0477e367f0d4947e430a/pkg/sql/opt/xform/testdata/rules/select#L1735)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] an partial -> a partial
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 4 of 4 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60553)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

mgartner commented:
bors r+
